pkg/csi/service/node.go

```diff
@@ -327,7 +333,98 @@ func (s *service) NodeGetVolumeStats(
 	req *csi.NodeGetVolumeStatsRequest) (
 	*csi.NodeGetVolumeStatsResponse, error) {
 
-	return nil, nil
+	var err error
+	targetPath := req.GetVolumePath()
+	if targetPath == "" {
+		err = fmt.Errorf("targetpath %v is empty", targetPath)
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	dev, err := getDevFromMount(targetPath)
+	if err != nil {
+		err = fmt.Errorf("unable to get targetpath %v device", targetPath)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if dev == nil {
+		err = fmt.Errorf("could not find device mounted on targetpath %v", targetPath)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	//TODO Check that the matching device is a vSphere volume, and that the volID matches the mount point
+
+	volMetrics, err := getMetrics(targetPath)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	available, ok := (*(volMetrics.Available)).AsInt64()
+	if !ok {
+		klog.Errorf("failed to fetch available bytes")
```

Comment:
Does it make sense to return if here `!ok`?

Reply:
According to the CSI Spec, 

> If the volume is being used in BlockVolume mode then used and available MAY be omitted from usage field of NodeGetVolumeStatsResponse. Similarly, inode information MAY be omitted from NodeGetVolumeStatsResponse when unavailable. 

So it seems that only the `capacity` field is mandatory, at least in `BlockVolume` mode.
If you prefer, we can make all fields mandatory by returning `nil` for each field if we can't get the metric.

Reply:
If that's what the spec says, then let's keep it as is!